components/infra-proxy-service/server/util_test.go

```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
 	assert.Equal(t, []string{}, SafeSliceFromMap(testData, "test3"))
 	assert.Equal(t, []string{}, SafeSliceFromMap(testData, "no-key"))
 }
+
+func TestSubstractSlice(t *testing.T) {
```

Comment:
Add tests:
* subtract both tag3 and unknown from the list of tag1, tag2, and tag3.
* subtract empty list from a non-empty list
* subtract a non-empty list from an empty list
* subtract whole list from itself

Reply:
Thank you for providing much detailed test cases scenario!